test/e2e/apimachinery/apiserver_identity.go

```diff
@@ -0,0 +1,169 @@
+	admissionapi "k8s.io/pod-security-admission/api"
+)
+
+func getControlPlaneHostname(node *v1.Node) (string, error) {
+	nodeAddresses := e2enode.GetAddresses(node, v1.NodeExternalIP)
+	if len(nodeAddresses) == 0 {
+		return "", errors.New("no valid addresses to use for SSH")
+	}
+
+	controlPlaneAddress := nodeAddresses[0]
+
+	host := controlPlaneAddress + ":" + e2essh.SSHPort
+	result, err := e2essh.SSH("hostname", host, framework.TestContext.Provider)
+	if err != nil {
+		return "", err
+	}
+
+	if result.Code != 0 {
+		return "", fmt.Errorf("encountered non-zero exit code when running hostname command: %d", result.Code)
+	}
+
+	return result.Stdout, nil
+}
+
+// restartAPIServer attempts to restart the kube-apiserver on a node
+func restartAPIServer(node *v1.Node) error {
+	nodeAddresses := e2enode.GetAddresses(node, v1.NodeExternalIP)
+	if len(nodeAddresses) == 0 {
+		return errors.New("no valid addresses to use for SSH")
+	}
+
+	controlPlaneAddress := nodeAddresses[0]
+	cmd := "pidof kube-apiserver | xargs sudo kill"
+	framework.Logf("Restarting kube-apiserver via ssh, running: %v", cmd)
+	result, err := e2essh.SSH(cmd, net.JoinHostPort(controlPlaneAddress, e2essh.SSHPort), framework.TestContext.Provider)
+	if err != nil || result.Code != 0 {
+		e2essh.LogResult(result)
+		return fmt.Errorf("couldn't restart controller-manager: %v", err)
+	}
+	return nil
+}
+
+// This test requires that --feature-gates=APIServerIdentity=true be set on the apiserver
+var _ = SIGDescribe("kube-apiserver identity [Feature:APIServerIdentity]", func() {
+	f := framework.NewDefaultFramework("kube-apiserver-identity")
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelPrivileged
+
+	ginkgo.It("kube-apiserver identity should persist after restart [Disruptive]", func() {
+		e2eskipper.SkipUnlessProviderIs("gce")
+
+		client := f.ClientSet
+
+		var controlPlaneNodes []v1.Node
+		nodes, err := client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+		framework.ExpectNoError(err)
+
+		for _, node := range nodes.Items {
+			if _, ok := node.Labels["node-role.kubernetes.io/control-plane"]; ok {
+				controlPlaneNodes = append(controlPlaneNodes, node)
+				continue
+			}
+
+			if _, ok := node.Labels["node-role.kubernetes.io/master"]; ok {
+				controlPlaneNodes = append(controlPlaneNodes, node)
+				continue
+			}
+
+			for _, taint := range node.Spec.Taints {
+				if taint.Key == "node-role.kubernetes.io/master" {
+					controlPlaneNodes = append(controlPlaneNodes, node)
+					break
+				}
+
+				if taint.Key == "node-role.kubernetes.io/control-plane" {
+					controlPlaneNodes = append(controlPlaneNodes, node)
+					break
+				}
+			}
+		}
+
+		leases, err := client.CoordinationV1().Leases(metav1.NamespaceSystem).List(context.TODO(), metav1.ListOptions{
+			LabelSelector: "k8s.io/component=kube-apiserver",
+		})
+		framework.ExpectNoError(err)
+		framework.ExpectEqual(len(leases.Items), len(controlPlaneNodes), "unexpected number of leases")
```

Comment:
Can this flake in CI?

Reply:
Only if you run it on a cluster while it is still being provisioned. Otherwise the node check above should prevent flakes

Reply:
I guess it doesn't hurt to add a retry loop here